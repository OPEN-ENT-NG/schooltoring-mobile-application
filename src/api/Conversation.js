import axios from "axios";
import EventTracker from "../api/EventTracker";

async function getConversations() {
  try {
    const request = {
      method: "GET",
      url: `${global.config.auth.endpoint}/schooltoring/conversations`
    };
    const { data } = await axios(request);
    return data;
  } catch (err) {
    throw err;
  }
}

async function getMessages(requestId, page) {
  let pageParameter = page ? `?page=${page}` : "";
  try {
    const request = {
      method: "GET",
      url: `${
        global.config.auth.endpoint
      }/schooltoring/conversation/${requestId}/messages${pageParameter}`
    };
    const { data } = await axios(request);
    return data;
  } catch (err) {
    throw err;
  }
}

async function postMessage(requestId, message) {
  try {
    const request = {
      method: "POST",
      url: `${
        global.config.auth.endpoint
      }/schooltoring/conversation/${requestId}/message`,
      data: message
    };
    await axios(request);
    EventTracker.trackEvent(
      EventTracker.events.MESSAGE.SEND,
      EventTracker.category.MESSAGE
    );
  } catch (err) {
    throw err;
  }
}

export default {
  getConversations,
  getMessages,
  postMessage
};
