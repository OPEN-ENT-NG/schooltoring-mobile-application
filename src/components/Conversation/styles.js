import { StyleSheet } from "react-native";

import { COLORS } from "../../styles/common";

const textColor = {
  color: COLORS.TEXT
};

export default StyleSheet.create({
  conversation: {
    flexDirection: "row",
    paddingHorizontal: 10
  },
  avatar: {
    alignSelf: "flex-start",
    flex: 1,
    alignItems: "center",
    paddingVertical: 20,
    marginHorizontal: 5
  },
  content: {
    flex: 4,
    justifyContent: "space-evenly",
    borderBottomColor: "#000",
    borderBottomWidth: StyleSheet.hairlineWidth
  },
  name: {
    ...textColor,
    fontWeight: "bold",
    fontSize: 20
  },
  message: {
    flexDirection: "row",
    justifyContent: "space-evenly"
  },
  icon: {
    fontSize: 20,
    marginRight: 10
  },
  text: {
    ...textColor,
    flex: 5,
    fontSize: 16,
    flexWrap: "wrap"
  },
  date: {
    flexGrow: 2,
    alignItems: "flex-end"
  },
  dateText: {
    ...textColor
  }
});
