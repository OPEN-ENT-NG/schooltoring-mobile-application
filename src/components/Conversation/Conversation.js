import React from "react";
import { View, Text } from "react-native";
import PropTypes from "prop-types";

import Icon from "react-native-vector-icons/MaterialIcons";

import Touchable from "../../components/Touchable/Touchable";
import styles from "./styles";
import { COLORS } from "../../styles/common";

import Avatar from "../Avatar/Avatar";

const getColor = state => {
  return state === "STRENGTH" ? COLORS.SECONDARY : COLORS.PRIMARY;
};

const Conversation = props => {
  return (
    <Touchable onPress={props.onPress}>
      <View style={styles.conversation}>
        <View style={styles.avatar}>
          <Avatar
            size={60}
            src={`${global.config.auth.endpoint}${props.userinfo.avatar}`}
          />
        </View>
        <View style={styles.content}>
          <Text style={styles.name} id="username-field">
            {props.userinfo.username}
          </Text>

          <View style={styles.message}>
            <Icon
              name="chat"
              color={getColor(props.state)}
              style={styles.icon}
            />
            <Text style={styles.text} numberOfLines={1} id="message-field">
              {props.message}
            </Text>

            <View style={styles.date}>
              <Text style={styles.dateText}>{props.date}</Text>
            </View>
          </View>
        </View>
      </View>
    </Touchable>
  );
};

Conversation.propTypes = {
  state: PropTypes.string.isRequired,
  userinfo: PropTypes.shape({
    username: PropTypes.string.isRequired,
    avatar: PropTypes.string.isRequired
  }).isRequired,
  date: PropTypes.string.isRequired,
  message: PropTypes.string,
  onPress: PropTypes.func.isRequired
};

export default Conversation;
