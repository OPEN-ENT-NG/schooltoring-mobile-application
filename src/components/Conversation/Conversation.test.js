import React from "react";
import { shallow } from "enzyme";
import moment from "moment";

import state from "../../jest/state.json";
import Conversation from "./Conversation";

const conversation = {
  userinfo: state.userinfo,
  state: "STRENGTH",
  date: "2018-08-20T12:35:24+00:00",
  message: "Test message 1"
};

const onPress = jest.fn();

const rendered = shallow(
  <Conversation
    state={conversation.state}
    userinfo={conversation.userinfo}
    date={moment(conversation.date).format("DD MMM Y")}
    message={conversation.message}
    onPress={onPress}
  />
);

describe("Render", () => {
  test("Conversation should render without crashing", () =>
    expect(rendered).toMatchSnapshot());

  test(`Username and message should be respectively '${
    conversation.userinfo.username
  }' and '${conversation.message}'`, () => {
    const usernameText = rendered.find("Text").find('[id="username-field"]');
    const messageText = rendered.find("Text").find('[id="message-field"]');

    expect(usernameText.contains(conversation.userinfo.username)).toBeTruthy();
    expect(messageText.contains(conversation.message)).toBeTruthy();
  });
});

describe("Interactions", () => {
  test("A press on conversation should call onPress prop function", () => {
    onPress.mockReset();
    const onPressButton = rendered.find("Touchable");
    onPressButton.props().onPress();

    expect(onPress).toHaveBeenCalled();
  });
});
