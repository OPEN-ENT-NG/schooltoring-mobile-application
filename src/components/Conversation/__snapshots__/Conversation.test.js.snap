// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render Conversation should render without crashing 1`] = `
<Touchable
  onPress={[MockFunction]}
>
  <Component
    style={
      Object {
        "flexDirection": "row",
        "paddingHorizontal": 10,
      }
    }
  >
    <Component
      style={
        Object {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "flex": 1,
          "marginHorizontal": 5,
          "paddingVertical": 20,
        }
      }
    >
      <Avatar
        size={60}
        src="http://ent3-core.gdapublic.fr/workspace/document/49a54de3-51f8-464b-81bb-a377f508f382"
      />
    </Component>
    <Component
      style={
        Object {
          "borderBottomColor": "#000",
          "borderBottomWidth": 0.5,
          "flex": 4,
          "justifyContent": "space-evenly",
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        id="username-field"
        style={
          Object {
            "color": "#232323",
            "fontSize": 20,
            "fontWeight": "bold",
          }
        }
      >
        John DOE
      </Text>
      <Component
        style={
          Object {
            "flexDirection": "row",
            "justifyContent": "space-evenly",
          }
        }
      >
        <Icon
          allowFontScaling={false}
          color="#e91e63"
          name="chat"
          size={12}
          style={
            Object {
              "fontSize": 20,
              "marginRight": 10,
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          id="message-field"
          numberOfLines={1}
          style={
            Object {
              "color": "#232323",
              "flex": 5,
              "flexWrap": "wrap",
              "fontSize": 16,
            }
          }
        >
          Test message 1
        </Text>
        <Component
          style={
            Object {
              "alignItems": "flex-end",
              "flexGrow": 2,
            }
          }
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#232323",
              }
            }
          >
            20 Aug 2018
          </Text>
        </Component>
      </Component>
    </Component>
  </Component>
</Touchable>
`;
